File: leetcode/1593-split-a-string-into-the-max-number-of-unique-substrings/Solution.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    private final Set<String> seen = new HashSet<>();
    private int best = 0;
    private String s;

    public int maxUniqueSplit(String s) {
        this.s = s;
        backtrack(0);
        return best;
    }

    private void backtrack(int start) {
        int n = s.length();
        if (start == n) {
            best = Math.max(best, seen.size());
            return;
        }
        if (seen.size() + (n - start) <= best) {
            return;
        }
        for (int end = start + 1; end <= n; end++) {
            String piece = s.substring(start, end);
            if (!seen.contains(piece)) {
                seen.add(piece);
                backtrack(end);
                seen.remove(piece);
            }
        }
    }
}

File: leetcode/1593-split-a-string-into-the-max-number-of-unique-substrings/solution.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        seen = set()
        n = len(s)
        best = 0

        def backtrack(start):
            nonlocal best
            if start == n:
                best = max(best, len(seen))
                return
            # prune: even taking every remaining char as its own piece
            # couldn't beat the best found so far
            if len(seen) + (n - start) <= best:
                return
            for end in range(start + 1, n + 1):
                piece = s[start:end]
                if piece not in seen:
                    seen.add(piece)
                    backtrack(end)
                    seen.remove(piece)

        backtrack(0)
        return best
